---
title: "Stool Microbiome Analysis"
author: "Kristen James"
date: "11/5/2021"
output:
  html_document:
    toc: true
    toc_float: true
---

# About

**Project:** FL100

**About:** Preproccess the data by determining the best prevalence thresholds to use. Run "standard" microbiome analysis focusing on plasma TMAO. This includes alpha and beta diversity, ANCOM differential abundance, and DESeq2 differential abundance analyses. 

**Inputs:** Inputs include: 

1) TMAO phyloseq object. Recall, the data from Dr. Kable was filtered to taxa with 2% abundance, which removes technical noise. Therefore, the remaining data is believable. Further filtering we do will be driven by our interest in more prevalent taxa. Furthermore, we should not recalculate the alpha diversity statistics after our filtering because that would incorrectly describe the microbiome environment.
2) Master phenotype data

**Outputs:** Outputs include: 

1) Plots per section [expand as analysis goes final]
2) Tables [expand as analysis goes final]

**Sources:**

1) [Example using Negative Binomial in Microbiome Differential Abundance Testing](https://bioconductor.org/packages/devel/bioc/vignettes/phyloseq/inst/doc/phyloseq-mixture-models.html)
2) [Workflow for Microbiome Data Analysis: from raw reads to community analyses](https://bioconductor.org/help/course-materials/2017/BioC2017/Day1/Workshops/Microbiome/MicrobiomeWorkflowII.html#different_ordination_projections)
3) [Example of what descriptive microbiome stats to report in the results](https://pubmed.ncbi.nlm.nih.gov/28407784/)


# Set Up

```{r, warning=FALSE, message=FALSE}
library(DESeq2)
library(phyloseq)
library(ggplot2)
library(dplyr)
library(vegan)
```

```{r Load phyloseq object}
load("../../data/processed/microbiome/phyloseq_objects/PSOtmao_211108.RData") #PSOtmao
```

```{r Add meta data to phyloseq object}
# Add age_cat
sample_data(PSOtmao)$age_cat <- ifelse(sample_data(PSOtmao)$age < 34, 1,
                         ifelse(sample_data(PSOtmao)$age >= 34 & sample_data(PSOtmao)$age < 50, 2, 3))
sample_data(PSOtmao)$age_cat <- as.factor(sample_data(PSOtmao)$age_cat)
# Add bmi_cat
sample_data(PSOtmao)$bmi_cat <- ifelse(sample_data(PSOtmao)$bmi_final.x < 25, 1,
                         ifelse(sample_data(PSOtmao)$bmi_final.x >= 25 & sample_data(PSOtmao)$bmi_final.x < 30, 2, 3))
sample_data(PSOtmao)$bmi_cat <- as.factor(sample_data(PSOtmao)$bmi_cat)

# Factor sex, BMI okay
sample_data(PSOtmao)$sex = factor(sample_data(PSOtmao)$sex)

# Make sure "below" is the reference factor
str(sample_data(PSOtmao)$tmao_mdn)
sample_data(PSOtmao)$tmao_mdn <- relevel(sample_data(PSOtmao)$tmao_mdn, ref = "below")

# Add TMAO quantiles
tmao_quantile <- quantile(sample_data(PSOtmao)$tmao)
sample_data(PSOtmao)$tmao_quantile <- ifelse(sample_data(PSOtmao)$tmao <= tmao_quantile[2], "Quantile1",
                             ifelse(sample_data(PSOtmao)$tmao > tmao_quantile[2] & sample_data(PSOtmao)$tmao <= tmao_quantile[3], "Quantile2",
                                   ifelse(sample_data(PSOtmao)$tmao > tmao_quantile[3] & sample_data(PSOtmao)$tmao <= tmao_quantile[4], "Quantile3", "Quantile4")))

# Factor it
sample_data(PSOtmao)$tmao_quantile = factor(sample_data(PSOtmao)$tmao_quantile)

# Add TMAO tertiles
tmao_tertile <- quantile(sample_data(PSOtmao)$tmao, c(0:3/3))
sample_data(PSOtmao)$tmao_tertile <- ifelse(sample_data(PSOtmao)$tmao <= tmao_tertile[2], "Tertile1",
                             ifelse(sample_data(PSOtmao)$tmao > tmao_tertile[2] & sample_data(PSOtmao)$tmao <= tmao_tertile[3], "Tertile2", "Tertile3"))
# Factor it
sample_data(PSOtmao)$tmao_tertile = factor(sample_data(PSOtmao)$tmao_tertile)
# Lookss good
table(sample_data(PSOtmao)$tmao_tertile)
```

# Preproccess

Table of read counts at the Phylum level. We will filter low abundant Phyla. We will also require that taxa are present in at least 10% of samples (at least 36 samples). We will then agglomerate at the Genus level. 

```{r Getting Familiar with Taxa Distributions}
print("The number of different bacteria at each phyla:")
table(tax_table(PSOtmao)[, "Phylum"])

# Prevalence - how many samples the taxa is observed in
prevdf = apply(X = otu_table(PSOtmao),
               MARGIN = ifelse(taxa_are_rows(PSOtmao), yes = 1, no = 2),
               FUN = function(x){sum(x > 0)})

prevdf = data.frame(Prevalence = prevdf,
                    TotalAbundance = taxa_sums(PSOtmao), # the total number of occurrences (even if only observed 50 times in 1 sample, the number would be 50)
                    tax_table(PSOtmao))

print("Table showing prevalence, total abundance, and taxonomy:")
prevdf[1:20,]

# Make table with mean prevalence, total prevalence
print("Table showing mean prevalencce and total prevalence per Phylum:")
plyr::ddply(prevdf, "Phylum", function(df_mb){cbind("Mean" = mean(df_mb$Prevalence),"Sum" = sum(df_mb$Prevalence))})

```
From the phylum table, we can see that p__Elusimicrobia, p__Spirochaetes, and p__Synergistetes have the same mean and sum, meaning they are only present in 1 sample. We are interested and confident in the more abundant taxa so manually filter them out. Per Dr. Kable's expert advice, she already filtered to 2% prevalence to remove technical error. We will increase the filtering to 5%, a common threshold in the literature. 

```{r Supervised Filtering 0.05}
# Define prevalence threshold as 5% of total samples
prevalenceThreshold <- 0.05 * nsamples(PSOtmao)
cat("The 10% prevalence threshold is", prevalenceThreshold, "samples.", "\n")

# Execute prevalence filter, using `prune_taxa()` function
keepTaxa = rownames(prevdf)[(prevdf$Prevalence >= prevalenceThreshold)]
PSOtmao1 = prune_taxa(keepTaxa, PSOtmao)

print("The number of different bacteria at each phyla:")
table(tax_table(PSOtmao1)[, "Phylum"])

# Look at phylum level distributions
ggplot(prevdf, aes(TotalAbundance, Prevalence / nsamples(PSOtmao1),color=Phylum)) +
  geom_hline(yintercept = 0.05, alpha = 0.5, linetype = 2) +  # Include a guess for parameter
  geom_point(size = 2, alpha = 0.7) +
  scale_x_log10() +  
  xlab("Total Abundance") + 
  ylab("Prevalence [Frac. Samples]") +
  facet_wrap(~Phylum) + 
  theme(legend.position="none")

# Agglomerate taxa
PSOtmao1G = tax_glom(PSOtmao1, "Genus", NArm = TRUE)
```

Great. Do your due diligence and inspect each phyloseq object to see how the filtering and glomming steps changed the number of taxonomy. 

```{r Phyloseq object inspection}
# Inspect the effect of each filtering step on the phyloseq object
PSOtmao
PSOtmao1
PSOtmao1G
```

Look at Genus level distributions of the bugs.

```{r Genus level distributions}
# Prevalence - how many samples the taxa is observed in
prevdf = apply(X = otu_table(PSOtmao1G),
               MARGIN = ifelse(taxa_are_rows(PSOtmao1G), yes = 1, no = 2),
               FUN = function(x){sum(x > 0)})

prevdf = data.frame(Prevalence = prevdf,
                    TotalAbundance = taxa_sums(PSOtmao1G), # the total number of occurrences (even if only observed 50 times in 1 sample, the number would be 50)
                    tax_table(PSOtmao1G))

print("Table showing prevalence, total abundance, and taxonomy:")
prevdf[1:10,]

# Family
ggplot(prevdf, aes(TotalAbundance, Prevalence / nsamples(PSOtmao1G),color=Family)) +
  geom_hline(yintercept = 0.05, alpha = 0.5, linetype = 2) + # Include a guess for parameter
  geom_point(size = 1, alpha = 0.7) +
  scale_x_log10() +
  xlab("Total Abundance") +
  ylab("Prevalence [Frac. Samples]") +
  facet_wrap(~Family) +
  theme(legend.position="none")

# Genus - have to blow it up to make it useful
ggplot(prevdf, aes(TotalAbundance, Prevalence / nsamples(PSOtmao1G),color=Genus)) +
  geom_hline(yintercept = 0.05, alpha = 0.5, linetype = 2) + # Include a guess for parameter
  geom_point(size = 1, alpha = 0.7) +
  scale_x_log10() +
  xlab("Total Abundance") +
  ylab("Prevalence [Frac. Samples]") +
  facet_wrap(~Genus) +
  theme(legend.position="none")
```

# Look for Outliers Using PCoA

We want to identify outlier taxa that might drive the statistical analyses. One way to do this is to look for outliers in the PCoA plot.

```{r PCoA for outliers}
PCoA_BC <- ordinate(PSOtmao1G, "PCoA", "bray")

# BC
# taxa is similar to species
taxa_plot <- plot_ordination(PSOtmao1G, PCoA_BC, type = "taxa")
taxa_plot
# It looks like there is an outlier in the top left corner of the plot. 
# Identify which taxa that bacteria is from and remove it. 
tp_data <- taxa_plot$data
plot(tp_data$Axis.2) # yes an outlier!
# Who is it?
outlier <- tp_data[tp_data$Axis.2 == max(tp_data$Axis.2), ] 
cat("Which genus is the outlier?", outlier$Genus, "\n")
```

Super interesting! The "outlier" is prevotella, which is interesting in the microbiome community as a marker of individuals who consume/live/have non-industrialized diets/environments/guts. We won't remove this bug afterall. 

Save

```{r save PSOtmao4G}
save(PSOtmao1, file = "../../data/processed/microbiome/phyloseq_objects/PSOtmao1_211118.RData")
save(PSOtmao1G, file = "../../data/processed/microbiome/phyloseq_objects/PSOtmao1G_211118.RData")
```

# Phylogenetic Abundance plots

## Family Plot of TMAO Quantile 

```{r Phylo Abundance Plot by TMAO Quantile}
# Agglomerate taxa
PSOtmao1F = tax_glom(PSOtmao1, "Family", NArm = TRUE)
PSOtmao1F.top10 <- microbiome::aggregate_top_taxa(PSOtmao1F, "Family", top = 10)

ps1F.10.family.comp <- microbiome::transform(PSOtmao1F.top10, transform="compositional")

plot.composition.relAbun <- microbiome::plot_composition(ps1F.10.family.comp,
                                             sample.sort = "Description",
                                             x.label = "tmao_quantile",
                                             group_by = "tmao_quantile",
                                             average_by = "tmao_quantile") +
  theme_minimal() +
  scale_fill_brewer("Family", palette = "Paired") + 
  ggtitle("Relative Abundance at Family Level")

plot.composition.relAbun
```

## Family Plot of TMAO Tertile 

```{r Tertiles}
# Agglomerate taxa
PSOtmao1F = tax_glom(PSOtmao1, "Family", NArm = TRUE)
PSOtmao1F.top10 <- microbiome::aggregate_top_taxa(PSOtmao1F, "Family", top = 10)
ps1F.10.family.comp <- microbiome::transform(PSOtmao1F.top10, transform="compositional") #compositional

plot.composition.relAbun <- microbiome::plot_composition(ps1F.10.family.comp,
                                             sample.sort = "Description",
                                             x.label = "tmao_tertile",
                                             group_by = "tmao_tertile",
                                             average_by = "tmao_tertile") +
  theme_minimal() +
  upstartr::scale_y_percent() + 
  scale_fill_brewer("Family", palette = "Paired") + 
  labs(x= "TMAO Tertiles", y = "Relative Abundance (%)", 
       title = "Relative Abundance of Top 10 Prevalent Families")

plot.composition.relAbun

# Save
tiff("../../Outputs/plots/microbiome/RelAbund_Fam_TMAOTertiles_2111.tiff", width = 4, height = 4, units = 'in', res = 300)
plot.composition.relAbun
dev.off()
```


# DESeq2

From the DESeq2 [documentation](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html), "Note: In order to benefit from the default settings of the package, you should put the variable of interest at the end of the formula and make sure the control level is the first level."

From Dr. Kable's example, 

> DESeq2 Set Up
> Test the impact of fiber cluster, accounting for covariates, using the log ratio test
>
> dds.ratio.batcho = phyloseq_to_deseq2(ps.genus, ~ Batch + k4_cluster)
> This performs a likelihood ratio test, which determines if the increased likelihood of the data in the full model, where we include all terms, is more than expected if the terms excluded in the reduced model are really zero. 
>
> dds.ratio.batcho <- DESeq(dds.ratio.batcho, test="LRT", reduced = ~ Batch )
> 
> The p-value is based on the LRT (so the same for everything), but the contrasts give the right log2 fold change.
> 
> resultsNames(dds.ratio.batcho)
> res.1v2.batcho <- results(dds.ratio.batcho, contrast=c("k4_cluster", "2", "1"))

## Factored by TMAO Median 

```{r DESeq2 analysis - TMAO median}
#~~~~~~~~~~~~~~~
# DESeq analysis
# TMAO median
#~~~~~~~~~~~~~~~
mb <- phyloseq_to_deseq2(PSOtmao1G, ~ sex + age + bmi_final.x + tmao_mdn)
mb.ratio <- DESeq(mb, test = "LRT", reduced = ~ sex + age + bmi_final.x) # remove variable of interest here, TMAO
res.ratio <- results(mb.ratio)
alpha = 0.05
alpha.big = 0.5
sigtab = res.ratio[which(res.ratio$padj < alpha.big), ]
sigtab = cbind(as(sigtab, "data.frame"), as(tax_table(PSOtmao1G)[rownames(sigtab), ], "matrix"))
sigtab$family_genus <- paste0(sigtab$Family, sigtab$Genus)
head(sigtab)

```

No results when we look at TMAO median and control for age, sex, and BMI.

## Factored by TMAO Tertile 

```{r DESeq2 analysis - TMAO tertile}
# DESeq analysis - tmao_tertile
mb <- phyloseq_to_deseq2(PSOtmao1G, ~ sex + age + bmi_final.x + tmao_tertile)
mb.ratio <- DESeq(mb, test = "LRT", reduced = ~ sex + age + bmi_final.x) # remove variable of interest here, TMAO
res.ratio <- results(mb.ratio, contrast = c("tmao_tertile", "Tertile1", "Tertile3")) # Look at most different tertiles
alpha = 0.05
sigtab = res.ratio[which(res.ratio$padj < alpha), ]
sigtab = cbind(as(sigtab, "data.frame"), as(tax_table(PSOtmao1G)[rownames(sigtab), ], "matrix"))
sigtab$family_genus <- paste0(sigtab$Family, sigtab$Genus)
head(sigtab)

#~~~~~~~~~~~~~~~
# Plot Results
#~~~~~~~~~~~~~~~
resOTU <- rownames(sigtab)
otuData <- as.data.frame(otu_table(PSOtmao1G))
otuData_desRes <- otuData[rownames(otuData) %in% resOTU,]
#otuData_desRes <- as.data.frame(t(otuData_desRes))
# Get sample data groupings
sampleData <- as.data.frame(sample_data(PSOtmao1G))
sampleData <- subset(sampleData, select = c(tmao_mdn, tmao_quantile, tmao_tertile, tmao))
# Get tax table
taxaData <- as.data.frame(tax_table(PSOtmao1G))
taxaData_res <- taxaData[rownames(taxaData) %in% resOTU,]
taxaData_res$OrderFamilyGenus <- paste0(taxaData_res$Order, taxaData_res$Family, taxaData_res$Genus)

# Merge
ggData <- merge(taxaData_res, otuData_desRes, by = 0)
rownames(ggData) <-  ggData$OrderFamilyGenus
ggData <- ggData[,-(1:9)]
ggData <- as.data.frame(t(ggData))
ggData <- merge(ggData, sampleData, by = 0)
rownames(ggData) <- ggData[,"Row.names"]
ggData <- subset(ggData, select = -c(Row.names))

# Loop Plot
nres <- length(resOTU)
for (i in 1:nres) { 
    print(ggplot(ggData, aes(x = tmao_tertile, y = log(ggData[,i]))) +
            geom_boxplot()+
            ylab(colnames(ggData)[i])) 
}

#~~~~~~~~~~~~~~~
# Plot Results Fold Change
#~~~~~~~~~~~~~~~
sigtabgen = subset(sigtab, !is.na(Genus))
# Phylum order
x = tapply(sigtabgen$log2FoldChange, sigtabgen$Phylum, function(x) max(x))
x = sort(x, TRUE)
sigtabgen$Phylum = factor(as.character(sigtabgen$Phylum), levels=names(x))
# Genus order
x = tapply(sigtabgen$log2FoldChange, sigtabgen$Genus, function(x) max(x))
x = sort(x, TRUE)
sigtabgen$Genus = factor(as.character(sigtabgen$Genus), levels=names(x))
DESeqPlot <- ggplot(sigtabgen, aes(y=family_genus, x=log2FoldChange, color=Phylum)) + 
  geom_vline(xintercept = 0.0, color = "gray", size = 0.5) +
  geom_point(size=3) + 
  ylab("Taxonomy") +
  xlab("Log2 Fold Change") + 
  scale_color_brewer(palette = "Paired") +
  theme_minimal() +
  theme(axis.text = element_text(size = 12), # x and y axis text size
        axis.title = element_text(size = 12), # x and y label text size
        legend.text = element_text(size = 12), # legend text size
        legend.title = element_text(size = 12), # label legend text size
        axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5))

DESeqPlot

# Save
tiff("../../Outputs/plots/microbiome/DESeq2_Genus_TMAOTertile1v3_2112.tiff", width = 8, height = 3, units = 'in', res = 300)
DESeqPlot
dev.off()
```
Save

```{r save deseq2 tmao tertile results}
write.csv(sigtabgen, "../../Outputs/tables/Microbiome/DESeq2_PSOtmao1G_TMAOTertile.csv")
```

## Factored by TMAO Quantile 

```{r DESeq2 analysis - TMAO quantile}
# DESeq analysis - tmao_quantile
mb <- phyloseq_to_deseq2(PSOtmao1G, ~ sex + age + bmi_final.x + tmao_quantile)
mb.ratio <- DESeq(mb, test = "LRT", reduced = ~ sex + age + bmi_final.x) # remove variable of interest here, TMAO
res.ratio <- results(mb.ratio, contrast = c("tmao_quantile", "Quantile1", "Quantile4"))
alpha = 0.05
sigtab = res.ratio[which(res.ratio$padj < alpha), ]
sigtab = cbind(as(sigtab, "data.frame"), as(tax_table(PSOtmao1G)[rownames(sigtab), ], "matrix"))
sigtab$family_genus <- paste0(sigtab$Family, sigtab$Genus)
head(sigtab)

#~~~~~~~~~~~~~~~
# Plot Results
#~~~~~~~~~~~~~~~
resOTU <- rownames(sigtab)
otuData <- as.data.frame(otu_table(PSOtmao1G))
otuData_desRes <- otuData[rownames(otuData) %in% resOTU,]
#otuData_desRes <- as.data.frame(t(otuData_desRes))
# Get sample data groupings
sampleData <- as.data.frame(sample_data(PSOtmao1G))
sampleData <- subset(sampleData, select = c(tmao_mdn, tmao_quantile, tmao_tertile, tmao))
# Get tax table
taxaData <- as.data.frame(tax_table(PSOtmao1G))
taxaData_res <- taxaData[rownames(taxaData) %in% resOTU,]
taxaData_res$OrderFamilyGenus <- paste0(taxaData_res$Order, taxaData_res$Family, taxaData_res$Genus)

# Merge
ggData <- merge(taxaData_res, otuData_desRes, by = 0)
rownames(ggData) <-  ggData$OrderFamilyGenus
ggData <- ggData[,-(1:9)]
ggData <- as.data.frame(t(ggData))
ggData <- merge(ggData, sampleData, by = 0)
rownames(ggData) <- ggData[,"Row.names"]
ggData <- subset(ggData, select = -c(Row.names))

# Loop Plot
nres <- length(resOTU)
for (i in 1:nres) { 
    print(ggplot(ggData, aes(x = tmao_quantile, y = log(ggData[,i]))) +
            geom_boxplot()+
            ylab(colnames(ggData)[i])) 
}

#~~~~~~~~~~~~~~~
# Plot Results Fold Change
#~~~~~~~~~~~~~~~
sigtabgen = subset(sigtab, !is.na(Genus))
# Phylum order
x = tapply(sigtabgen$log2FoldChange, sigtabgen$Phylum, function(x) max(x))
x = sort(x, TRUE)
sigtabgen$Phylum = factor(as.character(sigtabgen$Phylum), levels=names(x))
# Genus order
x = tapply(sigtabgen$log2FoldChange, sigtabgen$Genus, function(x) max(x))
x = sort(x, TRUE)
sigtabgen$Genus = factor(as.character(sigtabgen$Genus), levels=names(x))
ggplot(sigtabgen, aes(y=family_genus, x=log2FoldChange, color=Phylum)) + 
  geom_vline(xintercept = 0.0, color = "gray", size = 0.5) +
  geom_point(size=3) + 
  ylab("Taxonomy") +
  xlab("Log2 Fold Change") + 
  scale_color_brewer(palette = "Dark2") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5))
```

# Alpha Diversity Analysis

We do not need to redo the alpha diversity calculations because the filtering we did was not to remove technical variation but instead to focus our analysis on the more common players (also so we can be more statistically confident using less scarce data). Note, if you did want to calculate alpha diversity, the command estimate_richness() can accomplish that for you. 

Check the distributions of the alpha diversity metrics and run the multiple linear regression models with covariates. 

```{r Original Alpha diversity stats}
# Load master pheno file
phen <- readRDS("../../data/processed/phenotype/Phenotypes_211108.rds")
phen <- subset(phen, select = -c(sex, age, age_cat, bmi_cat))

# Grab PSO sample data
df <- as(sample_data(PSOtmao1G), "data.frame")
# Merge
df2 <- merge(phen, df, by = 0)
row.names(df2) <- df2[,"Row.names"]
df2 <- subset(df2, select =-c(Row.names))

#~~~~~~~~~~~~~~~~
# Compute
#~~~~~~~~~~~~~~~~

# Check distribution
shapiro.test(df2$shannon) # not normal, proceed with caution
shapiro.test(log(df2$shannon))
# Run linear regression
summary(lm(tmao_log ~ shannon + sex*age + cystatinc_bd1, df2))

# Check distribution
shapiro.test(df2$faith_pd) # normal
# Run linear regression
summary(lm(tmao_log ~ faith_pd + sex*age + cystatinc_bd1, df2))

# Check distribution
shapiro.test(df2$pielou_e) # not normal, proceed with caution
# Run linear regression
summary(lm(tmao_log ~ pielou_e + sex*age + cystatinc_bd1, df2))

# Check distribution
shapiro.test(df2$observed_otus) # normal
# Run linear regression
summary(lm(tmao_log ~ observed_otus + sex*age + cystatinc_bd1, df2))
```

## Alpha Diversity Plots

Make plots that demonstrate the relationship between TMAO and the alpha-diversity metrics. 

```{r Original Alpha Diversity Plots}
# Shannon
y_variable <- "Shannon"
x_variable <- "Natural Log of TMAO (uM)"
mdn_cut <- log(median(df2$tmao.x))
# Plot
shannon_plot <- ggplot(df2, aes(x=tmao_log, y=shannon)) +
  geom_point(aes(shape=bmi_cat, color = age_cat)) +
  geom_vline(xintercept = mdn_cut, linetype = "dashed") +
  geom_smooth(method = "lm", se = FALSE, fullrange = F) +
  theme_minimal() + 
  ylab(y_variable) +
  xlab(x_variable) +
  labs(color = "Age bin", shape = "BMI bin") +
  scale_color_brewer(palette = "Dark2") +
  theme(axis.text = element_text(size = 12), # x and y axis text size
        axis.title = element_text(size = 12), # x and y label text size
        legend.text = element_text(size = 12), # legend text size
        legend.title = element_text(size = 12), # label legend text size
        axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5))
shannon_plot

# Faith's Phylogenetic Diversity
y_variable <- "Faith's Phylogenetic Diversity"
x_variable <- "Natural Log of TMAO (uM)"
mdn_cut <- log(median(df2$tmao.x))
# Plot
faith_plot <- ggplot(df2, aes(x=tmao_log, y=faith_pd)) +
  geom_point(aes(shape=bmi_cat, color = age_cat)) +
  geom_vline(xintercept = mdn_cut, linetype = "dashed") +
  geom_smooth(method = "lm", se = FALSE, fullrange = F) +
  theme_minimal() + 
  ylab(y_variable) +
  xlab(x_variable) +
  labs(color = "Age bin", shape = "BMI bin") +
  scale_color_brewer(palette = "Dark2") +
  theme(axis.text = element_text(size = 12), # x and y axis text size
        axis.title = element_text(size = 12), # x and y label text size
        legend.text = element_text(size = 12), # legend text size
        legend.title = element_text(size = 12), # label legend text size
        axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5))
faith_plot

# Pielou's Evenness
y_variable <- "Pielou's Evenness"
x_variable <- "Natural Log of TMAO (uM)"
mdn_cut <- log(median(df2$tmao.x))
# Plot
pielou_plot <- ggplot(df2, aes(x=tmao_log, y=pielou_e)) +
  geom_point(aes(shape=bmi_cat, color = age_cat)) +
  geom_vline(xintercept = mdn_cut, linetype = "dashed") +
  geom_smooth(method = "lm", se = FALSE, fullrange = F) +
  theme_minimal() + 
  ylab(y_variable) +
  xlab(x_variable) +
  labs(color = "Age bin", shape = "BMI bin") +
  scale_color_brewer(palette = "Dark2") +
  theme(axis.text = element_text(size = 12), # x and y axis text size
        axis.title = element_text(size = 12), # x and y label text size
        legend.text = element_text(size = 12), # legend text size
        legend.title = element_text(size = 12), # label legend text size
        axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5))
pielou_plot

# Observed OTUs
y_variable <- "Observed OTUs"
x_variable <- "Natural Log of TMAO (uM)"
mdn_cut <- log(median(df2$tmao.x))
# Plot
obsOTU_plot <- ggplot(df2, aes(x=tmao_log, y=observed_otus)) +
  geom_point(aes(shape=bmi_cat, color = age_cat)) +
  geom_vline(xintercept = mdn_cut, linetype = "dashed") +
  geom_smooth(method = "lm", se = FALSE, fullrange = F) +
  theme_minimal() + 
  ylab(y_variable) +
  xlab(x_variable) +
  labs(color = "Age bin", shape = "BMI bin") +
  scale_color_brewer(palette = "Dark2") +
  theme(axis.text = element_text(size = 12), # x and y axis text size
        axis.title = element_text(size = 12), # x and y label text size
        legend.text = element_text(size = 12), # legend text size
        legend.title = element_text(size = 12), # label legend text size
        axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5))
obsOTU_plot
```

Save

```{r save alpha diversity plots}
# Save
tiff("../../Outputs/plots/microbiome/AlphaDiversity_Shannon_TMAO_2112.tiff", width = 6, height = 4, units = 'in', res = 300)
shannon_plot
dev.off()
```

# Beta Diversity Analysis

```{r Beta diversity calculations}
# PCoA
PCoA_WU <- ordinate(PSOtmao1G, "PCoA", "wunifrac")
PCoA_UnWU <- ordinate(PSOtmao1G, "PCoA", "uunifrac")
PCoA_BC <- ordinate(PSOtmao1G, "PCoA", "bray")

# Calculate distance matrix
dismax_wu = phyloseq::distance(PSOtmao1G, method="wunifrac", type = "samples")
dismax_uwu = phyloseq::distance(PSOtmao1G, method="unifrac", type = "samples")
dismax_bc = phyloseq::distance(PSOtmao1G, method="bray", type = "samples")

```

Save

```{r save beta matrices}
save(PCoA_WU, file = "../../data/processed/microbiome/diversity/PSOtmao1G_PCoA_WU.RData")
save(PCoA_UnWU, file = "../../data/processed/microbiome/diversity/PSOtmao1G_PCoA_UnWU.RData")
save(PCoA_BC, file = "../../data/processed/microbiome/diversity/PSOtmao1G_PCoA_BC.RData")

save(dismax_wu, file = "../../data/processed/microbiome/diversity/PSOtmao1G_dismax_WU.RData")
save(dismax_uwu, file = "../../data/processed/microbiome/diversity/PSOtmao1G_dismax_UWU.RData")
save(dismax_bc, file = "../../data/processed/microbiome/diversity/PSOtmao1G_dismax_BC.RData")
```


```{r Beta diversity stats}
# Get sample_data
df <- as(sample_data(PSOtmao1G), "data.frame")
df$tmao_mdn <- relevel(df$tmao_mdn, ref = "below")

#~~~~~~~~~~~
# Weighted
#~~~~~~~~~~~
# Beta dispersion
bdis <- betadisper(dismax_wu, df$tmao_mdn)
permutest(bdis) # NS, good
boxplot(bdis, las = 2)

# ADONIS (perMANOVA) analysis
adonis2(dismax_wu ~ tmao_log + sex + age + bmi_final.y, data=df, method = "wunifrac", perm=999) # .025
#adonis2(tmao_log ~ dismax_wu + sex*age + cystatinc_bd1, data=df, method = "wunifrac", perm=999) 

#~~~~~~~~~~~
# Unweighted
#~~~~~~~~~~~
# Beta dispersion
bdis <- betadisper(dismax_uwu, df$tmao_mdn)
permutest(bdis) # NS, good
boxplot(bdis, las = 2)

# ADONIS (perMANOVA) analysis
adonis2(dismax_uwu ~ tmao_log + sex + age + bmi_final.y, data=df, method = "unifrac", perm=999) # .001
#adonis2(tmao_log ~ dismax_uwu + sex*age + cystatinc_bd1, data=df, method = "unifrac", perm=999) 

#~~~~~~~~~~~
# BrayCurtis
#~~~~~~~~~~~
# Beta dispersion
bdis <- betadisper(dismax_bc, df$tmao_mdn)
permutest(bdis) # NS, good
boxplot(bdis, las = 2)

# ADONIS (perMANOVA) analysis
adonis2(dismax_bc ~ tmao_log + sex + age + bmi_final.y, data=df, method = "bray", perm=999) # .017
#adonis2(tmao_log ~ dismax_bc + sex*age + cystatinc_bd1, data=df, method = "bray", perm=999) 

```

## Beta Diversity Plots 

```{r beta diversity plots}
# Weighted
betaplot_WU <- plot_ordination(PSOtmao1G, PCoA_WU, 
                    color="tmao_mdn",
                    shape = NULL,
                    axes = 1:2) +
  scale_color_brewer(palette = "Dark2", labels = c("Below", "Above")) +
  stat_ellipse(linetype = 1) +
  labs(color = "TMAO Median Threshold") +
  theme_minimal() + 
  theme(axis.text = element_text(size = 12), # x and y axis text size
        axis.title = element_text(size = 12), # x and y label text size
        legend.text = element_text(size = 12), # legend text size
        legend.title = element_text(size = 12), # label legend text size
        axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5))

betaplot_WU

# Unweighted
betaplot_UWU <- plot_ordination(PSOtmao1G, PCoA_UnWU, 
                    color="tmao_mdn",
                    shape = NULL,
                    axes = 1:2) +
  scale_color_brewer(palette = "Dark2", labels = c("Below", "Above")) +
  stat_ellipse(linetype = 1) +
  labs(color = "TMAO Median Threshold") +
  theme_minimal() + 
  theme(axis.text = element_text(size = 12), # x and y axis text size
        axis.title = element_text(size = 12), # x and y label text size
        legend.text = element_text(size = 12), # legend text size
        legend.title = element_text(size = 12), # label legend text size
        axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5))

betaplot_UWU

# Bray Curtis
betaplot_bc <- plot_ordination(PSOtmao1G, PCoA_BC, 
                    color="tmao_mdn",
                    shape = NULL,
                    axes = 1:2) +
  scale_color_brewer(palette = "Dark2", labels = c("Below", "Above")) +
  stat_ellipse(linetype = 1) +
  labs(color = "TMAO Median Threshold") +
  theme_minimal() + 
  theme(axis.text = element_text(size = 12), # x and y axis text size
        axis.title = element_text(size = 12), # x and y label text size
        legend.text = element_text(size = 12), # legend text size
        legend.title = element_text(size = 12), # label legend text size
        axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5))

betaplot_bc

```

Save

```{r save beta diversity plots}
# Save
tiff("../../Outputs/plots/microbiome/BetaDiversity_PCoAWU_TMAOmdn_2112.tiff", width = 6, height = 4, units = 'in', res = 300)
betaplot_WU
dev.off()
```

# ANCOM

## ANCOM2 no-covariate analysis

```{r ANCOM - genus level}
source("../../Rscripts/ancom_v2.1.R")

# OTU table
MicroData <- otu_table(PSOtmao1G)
MicroData <- as.data.frame(as.matrix(MicroData), stringsAsFactors = F)
# Add 1
#MicroData2 <- MicroData +1
#MicroData2 <- as.data.frame(as.matrix(t(MicroData2),stringsAsFactors = F))
#dim(MicroData2) # 103 x 98

# Load the phenotype data
#------------------------
# Perform ANCOM analysis
# Step 1: Data preprocessing
# OTU data
otu_data <- otu_table(PSOtmao1G)
otu_data <- as.data.frame(as.matrix(otu_data), stringsAsFactors = F)
otu_id <- rownames(otu_data)

# Metadata 
#df <- readRDS("../data_processed/mapping/Biocrates_mapping_BD1_210809.rds")
#df$p.Cresol.SO4_bd1_mdn <- plyr::revalue(df$p.Cresol.SO4_bd1_mdn, c("0" = "< Median", "1" = "> Median"))
# Subset
MetaData2 <- df2 %>% dplyr::select(c(tmao_mdn, age, sex, bmi_final))
MetaData2$Sample.ID <- rownames(MetaData2)


feature_table = otu_data; sample_var = "Sample.ID"; group_var = NULL
out_cut = 0.05; zero_cut = 0.95; lib_cut = 1000; neg_lb = FALSE # make zero_cut=0.95 to match criteria used for other metabolites
# Function
prepro = feature_table_pre_process(feature_table = feature_table,
                                   meta_data = MetaData2,
                                   sample_var = sample_var,
                                   group_var = NULL,
                                   out_cut = out_cut,
                                   zero_cut = zero_cut,
                                   lib_cut = lib_cut,
                                   neg_lb = neg_lb)

# Step 2: ANCOM
feature_table = prepro$feature_table 
meta_data = prepro$meta_data
struc_zero = prepro$structure_zeros

# Define variables for ANCOM
main_var = "tmao_mdn"; p_adj_method = "BH"; alpha = 0.05
adj_formula = NULL ; rand_formula = NULL
# Run ANCCOM
t_start = Sys.time()
res = ANCOM(feature_table = feature_table, 
            meta_data = meta_data, 
            struc_zero = struc_zero, 
            main_var = main_var, 
            p_adj_method = p_adj_method, 
            alpha = alpha, 
            adj_formula = adj_formula, 
            rand_formula = rand_formula)
t_end = Sys.time()
t_run = t_end - t_start 
t_run # around 5s

# Results
res$out
res$fig$data

# Step 3: Volcano Plot
# Number of taxa except structural zeros
n_taxa = ifelse(is.null(struc_zero), nrow(feature_table), sum(apply(struc_zero, 1, sum) == 0))
# Cutoff values for declaring differentially abundant taxa
cut_off = c(0.9 * (n_taxa -1), 0.8 * (n_taxa -1), 0.7 * (n_taxa -1), 0.6 * (n_taxa -1))
names(cut_off) = c("detected_0.9", "detected_0.8", "detected_0.7", "detected_0.6")

# Annotation data
dat_ann = data.frame(x = min(res$fig$data$x), y = cut_off["detected_0.7"], label = "W[0.7]")

fig = res$fig +  
  geom_hline(yintercept = cut_off["detected_0.7"], linetype = "dashed") + 
  geom_text(data = dat_ann, aes(x = x, y = y, label = label), 
            size = 4, vjust = -0.5, hjust = 0, color = "orange", parse = TRUE)
fig  

# ORGANIZE PLOT

TaxaData2 <- readRDS("../../data/processed/microbiome/phyloseq_inputs/merged-taxonomy-dada2-retrained.rds")
rownames(TaxaData2) <- TaxaData2$Feature.ID

ancomRes <- res$out

test <- merge(ancomRes, TaxaData2, by.x = "taxa_id", by.y = "Feature.ID", all.x = TRUE, all.y = FALSE)

# Make plot 
tempdata <- test[test$detected_0.6 == TRUE,] # 0.6 but can make 0.7
asv_id <- unique(tempdata$taxa_id)

MicroData4 <- feature_table[row.names(feature_table) %in% asv_id,]

# Rename to the taxonomy 
#MicroData5 <- as.data.frame(as.matrix(t(MicroData4)))
TaxaData3 <- TaxaData2 %>% dplyr::select(Order, Family, Genus)
# Make a unique naming column that incorporates multiple levels of taxonomy
TaxaData3$OrderFamilyGenus <- paste(TaxaData3$Order, TaxaData3$Family, TaxaData3$Genus)
rownames(TaxaData3) <- rownames(TaxaData2)
# Merge
MicroData5 <- merge(MicroData4, TaxaData3,
                    by=0, all.x = T, all.y = F)

row.names(MicroData5) <- MicroData5$OrderFamilyGenus

MicroData5 <- MicroData5 %>%  dplyr::select(-Order, -Family, -Genus, -OrderFamilyGenus, -Row.names)
MicroData5 <- as.data.frame(as.matrix(t(MicroData5)))

# Merge with the phenotype data 
MetaData3 <- MetaData2 %>% dplyr::select(tmao_mdn)

ggData <- merge(MetaData3, MicroData5, by =0)
row.names(ggData) <- ggData$Row.names
ggData <- ggData %>% dplyr::select(-Row.names)

# Gather the data
library(tidyr)
ggData2 <- gather(ggData,
                  key = "Taxa",
                  value = "Abundance",
                  -tmao_mdn)
                
ggData2$Abundance <- log(ggData2$Abundance)

# Plot it
p <- ggplot(data = ggData2,
            aes(x = Taxa, 
                y = Abundance,
                fill = tmao_mdn), 
            alpha = 0.1) +
  geom_boxplot(lwd=.25) +
  xlab(NULL) +
  ylab("Log of Taxa Abundance") +
  theme_bw() + # note order matters here - need bw to be before theme()
  theme(axis.text.y = element_text( hjust = 1),
        axis.text.x = element_text(angle = 60, hjust = 1),
        text=element_text(size=11))  +
  scale_fill_manual(values=c("#2D708EFF", "#B8DE29FF")) + 
  #scale_fill_viridis(discrete = TRUE) +
  #coord_flip() +
  labs(fill = "TMAO \n Classification")

p
```


## ANCOM2 covariate analysis

ANCOM2 lets you include covariates into the analysis. If we add the same covariates as in the DESeq2 analysis, do we get similar results? 

```{r ANCOM - genus level with covariates median}
source("../../Rscripts/ancom_v2.1.R")

# OTU table
MicroData <- otu_table(PSOtmao1G)
MicroData <- as.data.frame(as.matrix(MicroData), stringsAsFactors = F)
# Add 1
#MicroData2 <- MicroData +1
#MicroData2 <- as.data.frame(as.matrix(t(MicroData2),stringsAsFactors = F))
#dim(MicroData2) # 103 x 98

# Load the phenotype data
#------------------------
# Perform ANCOM analysis
# Step 1: Data preprocessing
# OTU data
otu_data <- otu_table(PSOtmao1G)
otu_data <- as.data.frame(as.matrix(otu_data), stringsAsFactors = F)
otu_id <- rownames(otu_data)

# Metadata 
#df <- readRDS("../data_processed/mapping/Biocrates_mapping_BD1_210809.rds")
#df$p.Cresol.SO4_bd1_mdn <- plyr::revalue(df$p.Cresol.SO4_bd1_mdn, c("0" = "< Median", "1" = "> Median"))
# Subset
MetaData2 <- df2 %>% dplyr::select(c(tmao_mdn, age, sex, bmi_final))
MetaData2$Sample.ID <- rownames(MetaData2)


feature_table = otu_data; sample_var = "Sample.ID"; group_var = NULL
out_cut = 0.05; zero_cut = 0.95; lib_cut = 1000; neg_lb = FALSE # make zero_cut=0.95 to match criteria used for other metabolites
# Function
prepro = feature_table_pre_process(feature_table = feature_table,
                                   meta_data = MetaData2,
                                   sample_var = sample_var,
                                   group_var = NULL,
                                   out_cut = out_cut,
                                   zero_cut = zero_cut,
                                   lib_cut = lib_cut,
                                   neg_lb = neg_lb)

# Step 2: ANCOM
feature_table = prepro$feature_table 
meta_data = prepro$meta_data
struc_zero = prepro$structure_zeros

# Define variables for ANCOM
main_var = "tmao_mdn"; p_adj_method = "BH"; alpha = 0.05
adj_formula = c("age + sex + bmi_final"); rand_formula = NULL
# Run ANCCOM
t_start = Sys.time()
res = ANCOM(feature_table = feature_table, 
            meta_data = meta_data, 
            struc_zero = struc_zero, 
            main_var = main_var, 
            p_adj_method = p_adj_method, 
            alpha = alpha, 
            adj_formula = adj_formula, 
            rand_formula = rand_formula)
t_end = Sys.time()
t_run = t_end - t_start 
t_run # around 5s

# Results
res$out
res$fig$data

# Step 3: Volcano Plot
# Number of taxa except structural zeros
n_taxa = ifelse(is.null(struc_zero), nrow(feature_table), sum(apply(struc_zero, 1, sum) == 0))
# Cutoff values for declaring differentially abundant taxa
cut_off = c(0.9 * (n_taxa -1), 0.8 * (n_taxa -1), 0.7 * (n_taxa -1), 0.6 * (n_taxa -1))
names(cut_off) = c("detected_0.9", "detected_0.8", "detected_0.7", "detected_0.6")

# Annotation data
dat_ann = data.frame(x = min(res$fig$data$x), y = cut_off["detected_0.7"], label = "W[0.7]")

fig = res$fig +  
  geom_hline(yintercept = cut_off["detected_0.7"], linetype = "dashed") + 
  geom_text(data = dat_ann, aes(x = x, y = y, label = label), 
            size = 4, vjust = -0.5, hjust = 0, color = "orange", parse = TRUE)
fig  

# ORGANIZE PLOT

TaxaData2 <- readRDS("../../data/processed/microbiome/phyloseq_inputs/merged-taxonomy-dada2-retrained.rds")
rownames(TaxaData2) <- TaxaData2$Feature.ID

ancomRes <- res$out

test <- merge(ancomRes, TaxaData2, by.x = "taxa_id", by.y = "Feature.ID", all.x = TRUE, all.y = FALSE)

# Make plot 
tempdata <- test[test$detected_0.6 == TRUE,] # 0.6 but can make 0.7
asv_id <- unique(tempdata$taxa_id)

MicroData4 <- feature_table[row.names(feature_table) %in% asv_id,]

# Rename to the taxonomy 
#MicroData5 <- as.data.frame(as.matrix(t(MicroData4)))
TaxaData3 <- TaxaData2 %>% dplyr::select(Order, Family, Genus)
# Make a unique naming column that incorporates multiple levels of taxonomy
TaxaData3$OrderFamilyGenus <- paste(TaxaData3$Order, TaxaData3$Family, TaxaData3$Genus)
rownames(TaxaData3) <- rownames(TaxaData2)
# Merge
MicroData5 <- merge(MicroData4, TaxaData3,
                    by=0, all.x = T, all.y = F)

row.names(MicroData5) <- MicroData5$OrderFamilyGenus

MicroData5 <- MicroData5 %>%  dplyr::select(-Order, -Family, -Genus, -OrderFamilyGenus, -Row.names)
MicroData5 <- as.data.frame(as.matrix(t(MicroData5)))

# Merge with the phenotype data 
MetaData3 <- MetaData2 %>% dplyr::select(tmao_mdn)

ggData <- merge(MetaData3, MicroData5, by =0)
row.names(ggData) <- ggData$Row.names
ggData <- ggData %>% dplyr::select(-Row.names)

# Gather the data
library(tidyr)
ggData2 <- gather(ggData,
                  key = "Taxa",
                  value = "Abundance",
                  -tmao_mdn)
                
ggData2$Abundance <- log(ggData2$Abundance)

# Plot it
p <- ggplot(data = ggData2,
            aes(x = Taxa, 
                y = Abundance,
                fill = tmao_mdn), 
            alpha = 0.1) +
  geom_boxplot(lwd=.25) +
  xlab(NULL) +
  ylab("Log of Taxa Abundance") +
  theme_bw() + # note order matters here - need bw to be before theme()
  theme(axis.text.y = element_text( hjust = 1),
        axis.text.x = element_text(angle = 60, hjust = 1),
        text=element_text(size=11))  +
  scale_fill_manual(values=c("#2D708EFF", "#B8DE29FF")) + 
  #scale_fill_viridis(discrete = TRUE) +
  #coord_flip() +
  labs(fill = "TMAO \n Classification")

p
```
# Stats for paper

Take a TMAO centric view point. So for the following questions, consider the whole group and the TMAO quantiles/tertiles. 

```{r Desc.Stats set up}
# Get sample data groupings
sampleData <- as.data.frame(sample_data(PSOtmao1))
sampleData <- subset(sampleData, select = c(tmao_mdn, tmao_quantile, tmao_tertile, tmao))

# Make lists of participants in each group
tertile1 <- rownames(sampleData[sampleData$tmao_tertile == "Tertile1",])
tertile2 <- rownames(sampleData[sampleData$tmao_tertile == "Tertile2",])
tertile3 <- rownames(sampleData[sampleData$tmao_tertile == "Tertile3",])

# Get OTU table
otuTbl_PSO1 <- as.data.frame(otu_table(PSOtmao1))
otuTbl_PSO1 <- as.data.frame(t(otuTbl_PSO1)) # subjects to rows

# Get tax table
taxaData <- as.data.frame(tax_table(PSOtmao1))
taxaData$PCOFG <- paste0(taxaData$Phylum, taxaData$Class, taxaData$Order, taxaData$Family, taxaData$Genus)

```

How many OTUs (not glommed) in analysis?

```{r How many}
#~~~~~~~~~~
# Tertile 1
#~~~~~~~~~~
# What is the average number of OTUs per participant in tertile 1?
taxa_occurrences_T1 <- sapply(otuTbl_PSO1[row.names(otuTbl_PSO1) %in% tertile1,], function(x) sum(x>0))
summary(taxa_occurrences_T1)

otuTbl_PSO1_smpCol <- as.data.frame(t(otuTbl_PSO1))
ppl_occurrences_T1 <- sapply(otuTbl_PSO1_smpCol[colnames(otuTbl_PSO1_smpCol) %in% tertile1,], function(x) sum(x>0))
summary(ppl_occurrences_T1)

#~~~~~~~~~~
# Tertile 2
#~~~~~~~~~~
# What is the average number of OTUs per participant in tertile 1?
taxa_occurrences_T2 <- sapply(otuTbl_PSO1[row.names(otuTbl_PSO1) %in% tertile2,], function(x) sum(x>0))
summary(taxa_occurrences_T2)

otuTbl_PSO1_smpCol <- as.data.frame(t(otuTbl_PSO1))
ppl_occurrences_T2 <- sapply(otuTbl_PSO1_smpCol[colnames(otuTbl_PSO1_smpCol) %in% tertile2,], function(x) sum(x>0))
summary(ppl_occurrences_T2)

#~~~~~~~~~~
# Tertile 3
#~~~~~~~~~~
# What is the average number of OTUs per participant in tertile 1?
taxa_occurrences_T3 <- sapply(otuTbl_PSO1[row.names(otuTbl_PSO1) %in% tertile3,], function(x) sum(x>0))
summary(taxa_occurrences_T3)

otuTbl_PSO1_smpCol <- as.data.frame(t(otuTbl_PSO1))
ppl_occurrences_T3 <- sapply(otuTbl_PSO1_smpCol[colnames(otuTbl_PSO1_smpCol) %in% tertile3,], function(x) sum(x>0))
summary(ppl_occurrences_T3)
```


How many OTUs are present in 5%, 25%, 50% of samples?

```{r}

```

What is the firmicutes to bacteroidetes ratio per tertile?
How common are Firmicutes? Mean, SD, range?
How common are Bacteroidetes? Mean, SD, range?

```{r Firmicutes to bacteroidetes ratio per tertile}
PSOtmao1_Phylum <- tax_glom(PSOtmao1, "Phylum", NArm = TRUE)
PSOtmao1_Phylum_tertile1 <- subset_samples(PSOtmao1_Phylum, tmao_tertile == "Tertile1")
PSOtmao1_Phylum_tertile2 <- subset_samples(PSOtmao1_Phylum, tmao_tertile == "Tertile2")
PSOtmao1_Phylum_tertile3 <- subset_samples(PSOtmao1_Phylum, tmao_tertile == "Tertile3")

#~~~~~~~~~~
# Tertile 1
#~~~~~~~~~~
# Prevalence - how many samples the taxa is observed in
prevalence_Phy_tert1 = apply(X = otu_table(PSOtmao1_Phylum_tertile1),
               MARGIN = ifelse(taxa_are_rows(PSOtmao1_Phylum_tertile1), yes = 1, no = 2),
               FUN = function(x){sum(x > 0)})

prevalence_Phy_tert1 = data.frame(Prevalence = prevalence_Phy_tert1,
                    TotalAbundance = taxa_sums(PSOtmao1_Phylum_tertile1), # the total number of occurrences (even if only observed 50 times in 1 sample, the number would be 50)
                    tax_table(PSOtmao1_Phylum_tertile1))

# Firmicutes:Bacteroidetes Abundance Ratio
firm_bacter_t1 <- prevalence_Phy_tert1$TotalAbundance[prevalence_Phy_tert1$Phylum == " p__Firmicutes"] / prevalence_Phy_tert1$TotalAbundance[prevalence_Phy_tert1$Phylum == " p__Bacteroidetes"] 

#~~~~~~~~~~
# Tertile 2
#~~~~~~~~~~
# Prevalence - how many samples the taxa is observed in
prevalence_Phy_tert2 = apply(X = otu_table(PSOtmao1_Phylum_tertile2),
               MARGIN = ifelse(taxa_are_rows(PSOtmao1_Phylum_tertile2), yes = 1, no = 2),
               FUN = function(x){sum(x > 0)})

prevalence_Phy_tert2 = data.frame(Prevalence = prevalence_Phy_tert2,
                    TotalAbundance = taxa_sums(PSOtmao1_Phylum_tertile2), # the total number of occurrences (even if only observed 50 times in 1 sample, the number would be 50)
                    tax_table(PSOtmao1_Phylum_tertile2))

firm_bacter_t2 <- prevalence_Phy_tert2$TotalAbundance[prevalence_Phy_tert2$Phylum == " p__Firmicutes"] / prevalence_Phy_tert2$TotalAbundance[prevalence_Phy_tert2$Phylum == " p__Bacteroidetes"] 

#~~~~~~~~~~
# Tertile 3
#~~~~~~~~~~
# Prevalence - how many samples the taxa is observed in
prevalence_Phy_tert3 = apply(X = otu_table(PSOtmao1_Phylum_tertile3),
               MARGIN = ifelse(taxa_are_rows(PSOtmao1_Phylum_tertile3), yes = 1, no = 2),
               FUN = function(x){sum(x > 0)})

prevalence_Phy_tert3 = data.frame(Prevalence = prevalence_Phy_tert3,
                    TotalAbundance = taxa_sums(PSOtmao1_Phylum_tertile3), # the total number of occurrences (even if only observed 50 times in 1 sample, the number would be 50)
                    tax_table(PSOtmao1_Phylum_tertile3))

firm_bacter_t3 <- prevalence_Phy_tert3$TotalAbundance[prevalence_Phy_tert3$Phylum == " p__Firmicutes"] / prevalence_Phy_tert3$TotalAbundance[prevalence_Phy_tert3$Phylum == " p__Bacteroidetes"] 

# Compare ratios
cat("Firmicutes:Bacteroidetes ratio tertile 1:", firm_bacter_t1)
cat("Firmicutes:Bacteroidetes ratio tertile 2:", firm_bacter_t2)
cat("Firmicutes:Bacteroidetes ratio tertile 3:", firm_bacter_t3)

```

What is the % abundance of a given taxa? Get summary statistics per tertile. 

Range percentage of taxa per person = n occurrences of bug/total occurrences of all bugs * 100

```{r Firmicute summary}
#~~~~~~~~~~
# Tertile 1
#~~~~~~~~~~
trans <- transform_sample_counts(PSOtmao1_Phylum_tertile1, function(x) x / sum(x))
melted <- psmelt(trans)
subdf <- melted[melted$Phylum == " p__Firmicutes", ]
cat("Summmary of Firmicute abundance in tertile 1:")
summary(subdf$Abundance)

#~~~~~~~~~~
# Tertile 2
#~~~~~~~~~~
trans <- transform_sample_counts(PSOtmao1_Phylum_tertile2, function(x) x / sum(x))
melted <- psmelt(trans)
subdf <- melted[melted$Phylum == " p__Firmicutes", ]
cat("Summmary of Firmicute abundance in tertile 2:")
summary(subdf$Abundance)

#~~~~~~~~~~
# Tertile 3
#~~~~~~~~~~
trans <- transform_sample_counts(PSOtmao1_Phylum_tertile3, function(x) x / sum(x))
melted <- psmelt(trans)
subdf <- melted[melted$Phylum == " p__Firmicutes", ]
cat("Summmary of Firmicute abundance in tertile 3:")
summary(subdf$Abundance)

```

```{r Bacteroidetes summary}
#~~~~~~~~~~
# Tertile 1
#~~~~~~~~~~
trans <- transform_sample_counts(PSOtmao1_Phylum_tertile1, function(x) x / sum(x))
melted <- psmelt(trans)
subdf <- melted[melted$Phylum == " p__Bacteroidetes", ]
cat("Summmary of Bacteroidetes abundance in tertile 1:")
summary(subdf$Abundance)

#~~~~~~~~~~
# Tertile 2
#~~~~~~~~~~
trans <- transform_sample_counts(PSOtmao1_Phylum_tertile2, function(x) x / sum(x))
melted <- psmelt(trans)
subdf <- melted[melted$Phylum == " p__Bacteroidetes", ]
cat("Summmary of Bacteroidetes abundance in tertile 2:")
summary(subdf$Abundance)

#~~~~~~~~~~
# Tertile 3
#~~~~~~~~~~
trans <- transform_sample_counts(PSOtmao1_Phylum_tertile3, function(x) x / sum(x))
melted <- psmelt(trans)
subdf <- melted[melted$Phylum == " p__Bacteroidetes", ]
cat("Summmary of Bacteroidetes abundance in tertile 3:")
summary(subdf$Abundance)

```

Genus: What are the most common genus?
Use the PSOtmao1G, subset to low-tertile, make prevalence data frame, sort in descending order. Repeat for other tertiles.

```{r Genus Prevalence by TMAO Tertile}
# Subset by tertiles 
PSOtmao1_tertile1 <- subset_samples(PSOtmao1, tmao_tertile == "Tertile1")
PSOtmao1_tertile2 <- subset_samples(PSOtmao1, tmao_tertile == "Tertile2")
PSOtmao1_tertile3 <- subset_samples(PSOtmao1, tmao_tertile == "Tertile3")

# How many taxa are there per tertile?
cat("n taxa in tertile 1:", ntaxa(PSOtmao1_tertile1))
cat("n taxa in tertile 2:", ntaxa(PSOtmao1_tertile2))
cat("n taxa in tertile 3:", ntaxa(PSOtmao1_tertile3))

# Now subset at the genus level to identify the most common genus
# Subset by tertiles (Genus)
PSOtmao1G_tertile1 <- subset_samples(PSOtmao1G, tmao_tertile == "Tertile1")
PSOtmao1G_tertile2 <- subset_samples(PSOtmao1G, tmao_tertile == "Tertile2")
PSOtmao1G_tertile3 <- subset_samples(PSOtmao1G, tmao_tertile == "Tertile3")

#~~~~~~~~~~
# Tertile 1
#~~~~~~~~~~
# Prevalence - how many samples the taxa is observed in
prevalence_tert1 = apply(X = otu_table(PSOtmao1G_tertile1),
               MARGIN = ifelse(taxa_are_rows(PSOtmao1G_tertile1), yes = 1, no = 2),
               FUN = function(x){sum(x > 0)})

prevalence_tert1 = data.frame(Prevalence = prevalence_tert1,
                    TotalAbundance = taxa_sums(PSOtmao1G_tertile1), # the total number of occurrences (even if only observed 50 times in 1 sample, the number would be 50)
                    tax_table(PSOtmao1G_tertile1))


#~~~~~~~~~~
# Tertile 2
#~~~~~~~~~~
# Prevalence - how many samples the taxa is observed in
prevalence_tert2 = apply(X = otu_table(PSOtmao1G_tertile2),
               MARGIN = ifelse(taxa_are_rows(PSOtmao1G_tertile2), yes = 1, no = 2),
               FUN = function(x){sum(x > 0)})

prevalence_tert2 = data.frame(Prevalence = prevalence_tert2,
                    TotalAbundance = taxa_sums(PSOtmao1G_tertile2), # the total number of occurrences (even if only observed 50 times in 1 sample, the number would be 50)
                    tax_table(PSOtmao1G_tertile2))
#~~~~~~~~~~
# Tertile 3
#~~~~~~~~~~
# Prevalence - how many samples the taxa is observed in
prevalence_tert3 = apply(X = otu_table(PSOtmao1G_tertile3),
               MARGIN = ifelse(taxa_are_rows(PSOtmao1G_tertile3), yes = 1, no = 2),
               FUN = function(x){sum(x > 0)})

prevalence_tert3 = data.frame(Prevalence = prevalence_tert3,
                    TotalAbundance = taxa_sums(PSOtmao1G_tertile3), # the total number of occurrences (even if only observed 50 times in 1 sample, the number would be 50)
                    tax_table(PSOtmao1G_tertile3))
```

View and sort top 10
```{r View abundant genus top 10}
cat("Most abundant genus in T1:")
head(prevalence_tert1[order(prevalence_tert1$TotalAbundance, decreasing= TRUE),], n = 10)
cat("Most abundant genus in T2:")
head(prevalence_tert2[order(prevalence_tert2$TotalAbundance, decreasing= TRUE),], n = 10)
cat("Most abundant genus in T3:")
head(prevalence_tert3[order(prevalence_tert3$TotalAbundance, decreasing= TRUE),], n = 10)

```

Save

```{r Save tertile prevalence csv}
write.csv(prevalence_tert1, "../../Outputs/tables/Microbiome/Prevalence_Genus_Tertile1.csv")
write.csv(prevalence_tert2, "../../Outputs/tables/Microbiome/Prevalence_Genus_Tertile2.csv")
write.csv(prevalence_tert3, "../../Outputs/tables/Microbiome/Prevalence_Genus_Tertile3.csv")
```

# Correlate Fecal Microbiome with Phenos

Now that we identified microbiome characteristics that are relevant to TMAO, ask if they 

1) Improve the variance explained between TMAO and a given cardiometabolic marker
2) Correlate to cardiometabolic phenos

```{r Get the microbiome data ready to combine with pheno data}
da <- read.csv("../../Outputs/tables/Microbiome/DESeq2_PSOtmao1G_TMAOTertile.csv", row.names = 1)
resOTU <- rownames(da)
otuData <- as.data.frame(otu_table(PSOtmao1G))
otuData_desRes <- otuData[rownames(otuData) %in% resOTU,]
otuData_desRes <- as.data.frame(t(otuData_desRes))
# rename for ease downstream
names(otuData_desRes)[names(otuData_desRes) == "bb1b75f41ff9c9db1d1de41e8388eb52"] <- "f__Bacteroidaceae_g__Bacteroides"
names(otuData_desRes)[names(otuData_desRes) == "b180a2455b236c103cf0bfe620095736"] <- "f__Peptostreptococcaceae_g__"
names(otuData_desRes)[names(otuData_desRes) == "e1a2800b24cdf9779b28dc897cddb12a"] <- "f__Christensenellaceae_g__"
names(otuData_desRes)[names(otuData_desRes) == "8f5ceded1cd7e86b8271d3fab24d322a"] <- "f__Lachnospiraceae_g__Butyrivibrio"
```

Next chunk is borrowed from "Characteristics_Diet_Cardiometabolic"

```{r Load phenotype data}
phen <- readRDS("../../data/processed/phenotype/Phenotypes_211108.rds")

# Complete TMAO data only
phen <- phen[complete.cases(phen$tmao == TRUE),]

# Get rid of NA ethnicity category
phen$ethnicity <- droplevels(phen$ethnicity)

# Factor and recode sex
phen$sex <- as.factor(phen$sex)
phen$sex <- plyr::revalue(phen$sex, c("1" = "Male", "2" = "Female"))

# Make TMAO log variable
phen$tmao_log <- log(phen$tmao)

# Quantile
tmao_quantile <- quantile(phen$tmao)
phen$tmao_quantile <- ifelse(phen$tmao <= tmao_quantile[2], "Quantile1",
                             ifelse(phen$tmao > tmao_quantile[2] & phen$tmao <= tmao_quantile[3], "Quantile2",
                                   ifelse(phen$tmao > tmao_quantile[3] & phen$tmao <= tmao_quantile[4], "Quantile3", "Quantile4")))

# Check it
table(phen$tmao_quantile)
phen$tmao_quantile <- as.factor(phen$tmao_quantile)
str(phen$tmao_quantile)

# Tertile
tmao_tertile <- quantile(phen$tmao, c(0:3/3))
phen$tmao_tertile <- ifelse(phen$tmao <= tmao_tertile[2], "Tertile1",
                             ifelse(phen$tmao > tmao_tertile[2] & phen$tmao <= tmao_tertile[3], "Tertile2", "Tertile3"))
# Factor it
phen$tmao_tertile <- factor(phen$tmao_tertile)
table(phen$tmao_tertile)

# Checked all variables for normality via shapiro.test() in console. TG, and glc ere the only ones that didn't pass.
phen$glc_bd1_ln <- log(phen$glc_bd1)
phen$tg_bd1_ln <- log(phen$tg_bd1)
phen$crp_bd1_ln <- log(phen$crp_bd1)
phen$tnfa_bd1_ln <- log(phen$tnfa_bd1)
phen$il6_bd1_ln <- log(phen$il6_bd1)
```

Merge phenotype and microbiome data for linear regression analysis
```{r Merge pheno and microbiome tertile DESeq results}
phen <- merge(phen, otuData_desRes, by=0)
rownames(phen) <- phen[,"Row.names"]
phen <- subset(phen, select = -c(Row.names))
```

Run the regression!

f__Bacteroidaceae_g__Bacteroides + f__Peptostreptococcaceae_g__ + f__Christensenellaceae_g__ + f__Lachnospiraceae_g__Butyrivibrio

```{r Cardiometabolic + Microbiome Multiple Linear Regression}
# Transfored variable list
transformed_keep_cardio_tbl <- c("sex", "age","bmi_final", "ethnicity", "waistavg_cm", "sysbpavgv2", "diabpavgv2", "endo_rhi", "endo_al75", "hdl_bd1", "ldl_bd1", "chol_bd1", "tg_bd1_ln", "glc_bd1_ln","tnfa_bd1_ln", "il6_bd1_ln", "crp_bd1_ln", "cystatinc_bd1", "choline", "betaine", "carnitine","tmao_log", "f__Bacteroidaceae_g__Bacteroides", "f__Peptostreptococcaceae_g__", "f__Christensenellaceae_g__", "f__Lachnospiraceae_g__Butyrivibrio")

#~~~~~~~~
# TMAO  
#~~~~~~~~
# Make a smaller df from master (the old data)
phen_sub <- phen[, colnames(phen) %in% transformed_keep_cardio_tbl]
cvd_biomark <- subset(phen_sub, select = -c(sex, age, ethnicity, cystatinc_bd1, tmao_log, f__Bacteroidaceae_g__Bacteroides, f__Peptostreptococcaceae_g__, f__Christensenellaceae_g__, f__Lachnospiraceae_g__Butyrivibrio)) # change variable here
n <- ncol(cvd_biomark)

# LCMS method
# Run linear model
my_lm <- lapply(1:n, function(x) lm(cvd_biomark[,x] ~ tmao_log + sex + age + cystatinc_bd1, phen_sub )) # include kcal
#my_lm <- lapply(1:n, function(x) lm(cvd_biomark[,x] ~ ethnicity, phen_sub ))
#my_lm <- lapply(1:n, function(x) lm(cvd_biomark[,x] ~ cystatinc_bd1, phen_sub ))

# Organize results
my_lm2 <- sapply(my_lm, function(x){summary(x)$adj.r.squared})
my_lm2.5 <- sapply(my_lm, function(x){summary(x)$coefficients[2,1]}) # [CVD row, estimate] corresponds to CVD B 
my_lm2.6 <- sapply(my_lm, function(x){summary(x)$coefficients[2,2]}) # [CVD row, Std.Error] corresponds to CVD Std. Error 
my_lm2.7 <- sapply(my_lm, function(x){summary(x)$coefficients[2,3]}) # [CVD row, T value] corresponds to CVD t value
my_lm3 <- sapply(my_lm, function(x){summary(x)$coefficients[2,4]}) # [CVD row, P] corresponds to CVD P value
my_lm361_lcms_cvd <- cbind(my_lm2, my_lm2.5, my_lm2.6, my_lm2.7, my_lm3)
colnames(my_lm361_lcms_cvd) <- c("adj.r.squared", "estimate", "Std.Error", "T.value","P")
row.names(my_lm361_lcms_cvd) <- colnames(cvd_biomark)
res <- as.data.frame(my_lm361_lcms_cvd)

# BH correction
res$p.adj <- stats::p.adjust(res$P, method = "BH", n = n) 

#~~~~~~~~
# TMAO + Bug
#~~~~~~~~
# Make a smaller df from master (the old data)
phen_sub <- phen[, colnames(phen) %in% transformed_keep_cardio_tbl]
cvd_biomark <- subset(phen_sub, select = -c(sex, age, ethnicity, cystatinc_bd1, tmao_log, f__Bacteroidaceae_g__Bacteroides, f__Peptostreptococcaceae_g__, f__Christensenellaceae_g__, f__Lachnospiraceae_g__Butyrivibrio)) # change variable here
n <- ncol(cvd_biomark)

# LCMS method
# Run linear model
my_lm <- lapply(1:n, function(x) lm(cvd_biomark[,x] ~ tmao_log + sex + age + cystatinc_bd1 + f__Bacteroidaceae_g__Bacteroides + f__Peptostreptococcaceae_g__ + f__Christensenellaceae_g__ + f__Lachnospiraceae_g__Butyrivibrio, phen_sub )) # add bugs

# Organize results
my_lm2 <- sapply(my_lm, function(x){summary(x)$adj.r.squared})
my_lm2.5 <- sapply(my_lm, function(x){summary(x)$coefficients[2,1]}) # [CVD row, estimate] corresponds to CVD B 
my_lm2.6 <- sapply(my_lm, function(x){summary(x)$coefficients[2,2]}) # [CVD row, Std.Error] corresponds to CVD Std. Error 
my_lm2.7 <- sapply(my_lm, function(x){summary(x)$coefficients[2,3]}) # [CVD row, T value] corresponds to CVD t value
my_lm3 <- sapply(my_lm, function(x){summary(x)$coefficients[2,4]}) # [CVD row, P] corresponds to CVD P value
my_lm361_lcms_cvd_des <- cbind(my_lm2, my_lm2.5, my_lm2.6, my_lm2.7, my_lm3)
colnames(my_lm361_lcms_cvd_des) <- c("adj.r.squared", "estimate", "Std.Error", "T.value","P")
row.names(my_lm361_lcms_cvd_des) <- colnames(cvd_biomark)
res_des <- as.data.frame(my_lm361_lcms_cvd_des)

# BH correction
res_des$p.adj <- stats::p.adjust(res_des$P, method = "BH", n = n) 
```

The family christensenellaceae is widely inversely associated with BMI. A review is [here](https://bmcbiol.biomedcentral.com/articles/10.1186/s12915-019-0699-4).

```{r Christensenellaceae genus and BMI}
summary(lm(f__Christensenellaceae_g__ ~ bmi_final + sex + age, phen))
```

